### PR TITLE
[JENKINS-55813] Check user validity in impersonate

### PR DESCRIPTION
**Pull request by jvz**
This ensures that user accounts used for performing actions are valid in all call sites.

See [JENKINS-55813](https://issues.jenkins-ci.org/browse/JENKINS-55813).

I'm opening this initially to test out some security realm changes elsewhere to validate this as including tests here would require extensive mocking.

### Proposed changelog entries

* JENKINS-55813, Security realms that provide account status checks are now properly enforced in unattended code that executes on behalf of a user.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by MRamonLeon — APPROVED** [review]

File: core/src/main/java/hudson/model/User.java
Comment (jvz) [reply core/src/main/java/hudson/model/User.java:428]: It's not a big an oversight as you'd expect because every single implementation of `UserDetails` returns `true` for all four of these methods. I'm working on supporting LDAP and AD specific attributes that would fill in real data in a separate PR for those plugins.

**Review by timja — APPROVED** [review]

**Comment by daniel-beck** [comment]
This looks like something that needs an escape hatch, just in case.

**Comment by varyvol** [comment]
@daniel-beck why do you think so? It seems like @jvz will work on supporting LDAP and AD, so if those are saying what should happen with those users, why should it be escaped?

**Review by res0nance — APPROVED** [review]

**Comment by jeffret-b** [comment]
I need to get a chance to review this and check some related things when I can get the chance.

**Review by jeffret-b — CHANGES_REQUESTED** [review]
I tested the behavior and it works great. It's a nice simple enhancement to plug this in and get it to work right.

The change really needs an automated test, though. It shouldn't have to rely on behaviors or tests in plugins to verify its proper behavior. It shouldn't be too hard to create an extension of `HudsonPrivateSecurityRealm` that sets these properties differently as needed to test. In `UserTest`, there is already `IdStrategySpecifyingSecurityRealm` that does something similar. Or just extend `SecurityRealm`.

**Comment by jeffret-b** [comment]
Love the new test!

I'd also like to have one for API tokens. If possible it would be really nice to have one for SSH connections. Not sure how difficult that would be. These would really help with regressions and would clarify earlier questions I had.

**Comment by jvz** [comment]
@jeffret-b added some proper tests to verify those situations.

**Review by jeffret-b — APPROVED** [review]
Looks great!

**Comment by jvz** [comment]
Marking this on hold. After experimenting more with LDAP, I've come to the conclusion that these four `UserDetails` attributes are all supposed to be checked by the LDAP server (or equivalent) which will return authentication failure appropriately. These attributes seem to be specific to Spring Security when using a JDBC store for authentication which Jenkins does not support. The Jenkins user database itself does not store metadata for those attributes either, so there's no reason to even check these attributes.

I'm going to try to confirm this hypothesis with more directory servers. One piece of supporting evidence to this whole situation is that I have yet to find an open source project in the wild that checks these attributes of an LDAP account other than LDAP servers themselves. For example, looking for the [policy code in Apache LDAP](https://github.com/apache/directory-server/search?q=in%3Afile+PasswordPolicyConfiguration), it seems as though this is automatically checked as part of a default authentication interceptor configuration. Looking at the [OpenLDAP overlay](https://git.openldap.org/openldap/openldap/-/blob/master/servers/slapd/overlays/ppolicy.c), it seems that OpenLDAP might be doing the same check only in bind operations. While it's possible there might be a way to look up the ppolicy attributes for accounts, it seems as though that should only be done by the LDAP server itself.

**Comment by jvz** [comment]
Re-opening this as it's still an improvement over the existing behavior. See https://github.com/jenkinsci/ldap-plugin/pull/50 for associated improvements.

**Review by rsandell — APPROVED** [review]

File: core/src/main/java/hudson/model/User.java
Comment (jeffret-b) [inline core/src/main/java/hudson/model/User.java:158]:
```suggestion
    private static boolean ALLOW_IMPERSONATION_OF_DISABLED_USERS = SystemProperties.getBoolean(User.class.getName() + ".allowImpersonationOfDisabledUsers");
```

**Review by jeffret-b — APPROVED** [review]

File: core/src/main/java/hudson/model/User.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/User.java:158]: Please create a downstream PR to https://www.jenkins.io/doc/book/managing/system-properties/

**Review by oleg-nenashev — APPROVED** [review]
The changelog entry is confusing. Security Plugin implementations can already supply the data. It is a Security Hardening change in Jenkins, because the Jenkins core starts consulting with this data provided by security realms. I suggest re-classifying it as RFE. 

It will also likely require re-testing with https://github.com/jenkinsci/jenkins/pull/4848 ,to ensure alignment with security realms.

File: test/src/test/java/hudson/model/UserTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/UserTest.java:35]: No wildcards please.

File: test/src/test/java/hudson/model/UserTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/UserTest.java:778]: You do not need this stub. Just use `org.acegisecurity.userdetails.User`. That will simplify the merge with #4848.

File: test/src/test/java/hudson/model/UserTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/UserTest.java:917]: It seems you are really testing behaviors of `ssh-cli-auth` and/or `sshd` here? Best to remove this part of the test, or leave it as an integration test in one of those modules, which will also be cleaner when I resurrect #3988 in simplified form.

File: core/src/main/java/hudson/model/User.java
Comment (jglick) [inline core/src/main/java/hudson/model/User.java:146]:
```suggestion
     * Jenkins does not allow users to be {@linkplain #impersonate(UserDetails)} impersonated} if its {@link UserDetails}
```

File: core/src/main/java/hudson/model/User.java
Comment (jglick) [inline core/src/main/java/hudson/model/User.java:150]:
```suggestion
     * {@link SecurityRealm} service or by the {@link AuthenticationManager}. Some realms return user details for these types
```

File: core/src/main/java/hudson/model/User.java
Comment (jglick) [reply core/src/main/java/hudson/model/User.java:158]: This is not scalable. We need to generate that sort of documentation from sources.

File: test/src/test/java/hudson/model/UserTest.java
Comment (jvz) [reply test/src/test/java/hudson/model/UserTest.java:917]: This test was added at @jeffret-b's request since the underlying issue related to API tokens and SSH keys. However, if I migrate this functionality all to the LDAP plugin specifically, then I could move these tests there.

Comment (jglick) [reply test/src/test/java/hudson/model/UserTest.java:917]: Or to `sshd-module`.

Comment (jvz) [reply test/src/test/java/hudson/model/UserTest.java:917]: Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

Comment (jvz) [reply test/src/test/java/hudson/model/UserTest.java:857]: I've added a comment in these two methods pointing out that the tests are common authentication mechanisms (you can see web login, API token login, and SSH login).

Comment (jglick) [reply test/src/test/java/hudson/model/UserTest.java:857]: > Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

Yes, which you can do temporarily using an incremental version, then switch to a weekly if and when this is released.

(Remember that there is no reason to hold back the `jenkins.version` in a `jenkins-module` since they are not distributed to users via the UC.)

Comment (jeffret-b) [reply test/src/test/java/hudson/model/UserTest.java:857]: API tokens are implemented in core so the test for it needs to be in core. This is probably as reasonable a place for it as anywhere. Putting the ssh test in sshd-module would be a better location.

Comment (jvz) [reply test/src/test/java/hudson/model/UserTest.java:857]: I was not aware about that regarding modules. Now that I'm a maintainer of sshd-module, that's not even too difficult to move this to. Good idea.

File: core/src/main/java/hudson/model/User.java
Comment (jglick) [inline core/src/main/java/hudson/model/User.java:445]: Still trying to figure out in https://github.com/jenkinsci/ldap-plugin/pull/50#issuecomment-707216245 why this should be done in core rather than the plugin defining the `SecurityRealm`.

Comment (jvz) [reply core/src/main/java/hudson/model/User.java:445]: Looking more closely at the code changes here, it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself since that will be checked both by `User.getUserDetailsForImpersonation()` and therefore by `User.impersonate()` as well as the few places in Core that call the `impersonate(UserDetails)` variant (those call `getUserDetailsForImpersonation()` first). Being that the `impersonate(UserDetails)` method is restricted from external use, it seems safe to move this downward.

Comment (jglick) [reply core/src/main/java/hudson/model/User.java:445]: > it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself

Right, that was my suggestion. The core PR might still be useful hardening in case there were some unknown realm out there which actually did return an invalid `Authentication`, but we would not be relying on the core check.

Comment (jvz) [reply core/src/main/java/hudson/model/User.java:445]: I've already reviewed all the `SecurityRealm` implementations for Jenkins in the wild (and proprietary ones) and found that none of them specify values other than `true` for these four properties. As I explained in greater detail in the LDAP plugin PR, it seems as though these properties are in the Spring Security API mostly for supporting [JDBC authentication providers](https://docs.spring.io/spring-security/site/docs/5.4.1/reference/html5/#servlet-authentication-jdbc) or other embedded user databases (e.g., Jenkins' own internal user database could optionally support these properties).

**Comment by timja** [comment]
it can't be on-hold and ready for merge.

@jeffret-b / @jvz what's the reason for on-hold? is it moving the test elsewhere or the system properties change? or something else?

**Comment by jvz** [comment]
> what's the reason for on-hold? is it moving the test elsewhere or the system properties change? or something else?

Two major things:

0. Wait for Spring Security upgrade to be merged first to avoid additional merge conflicts.
1. Move the attribute check logic into the underlying `UserDetailsService.loadUserByUsername()` used in specific security realms (like in the LDAP PR) in such a way to avoid any need to patch core.

I'll end up migrating some code from this PR into the others, but I think this PR is stalled until then.

File: test/src/test/java/hudson/model/UserTest.java
Comment (jvz) [reply test/src/test/java/hudson/model/UserTest.java:857]: Tests moved to https://github.com/jenkinsci/sshd-module/pull/39 which also discovered a bug in this test as it is.

File: core/src/main/java/hudson/model/User.java
Comment (jvz) [reply core/src/main/java/hudson/model/User.java:158]: https://github.com/jenkins-infra/jenkins.io/pull/3953

**Comment by jvz** [comment]
Based on further testing, it makes sense to include this check in core along with in the affected plugins. I've added a docs update for the system property as well.

**Comment by oleg-nenashev** [comment]
I do not see what blocks the merge here.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by jglick** [comment]
> it makes sense to include this check in core along with in the affected plugins

I see https://github.com/jenkinsci/ldap-plugin/pull/59 was merged, but I am still wondering why we would not just treat this as an implementation detail of `LDAPSecurityRealm`: if there is something wrong with the LDAP entry, throw an exception and refuse to authenticate, and ignore these parts of the Spring Security API as extraneous.

Anyway, no strong opinion, seems harmless to do it this way, just somewhat more complicated.

**Comment by jvz** [comment]
It's unexpected behavior that when a user says it's disabled, it is still allowed to use SSH keys or API tokens basically.

**Comment by jglick** [comment]
Agreed, I just thought that could be handled (along with GUI login) in `ldap-plugin` without this or https://github.com/jenkinsci/sshd-module/pull/39, by throwing authentication exceptions from suitable methods if the LDAP server indicates the user is disabled, rather than creating `Authentication` & `User` objects but setting various flags on them. Perhaps I misunderstood some subtlety.

**Comment by jvz** [comment]
The subtlety is due to API tokens being in core. Also, every other usage of `User.impersonate()` appears to expect a valid user, so this clears up previously ambiguous behavior. SSH keys can be handled in the sshd-module, but API tokens are core.

**Comment by jvz** [comment]
> I see [jenkinsci/ldap-plugin#59](https://github.com/jenkinsci/ldap-plugin/pull/59) was merged, but I am still wondering why we would not just treat this as an implementation detail of `LDAPSecurityRealm`: if there is something wrong with the LDAP entry, throw an exception and refuse to authenticate, and ignore these parts of the Spring Security API as extraneous.

Because `User.impersonate()` does _not_ call authenticate. It only looks up user details, and if I threw an error at this point, it will end up using `LegitimateButUnknownUserDetails` which would allow for bypassing the validity check!

**Comment by jglick** [comment]
> It only looks up user details, and if I threw an error at this point, it will end up using `LegitimateButUnknownUserDetails`

Only if the `UserDetailsService.loadUserByUsername` implementation throws `UserMayOrMayNotExistException2`. If you throw some other `AuthenticationException`, such as `AccountExpiredException`, from `LDAPUserDetailsService` instead of calling `setAccountNonExpired` and the like, then you will get the desired secure behavior AFAICT: `BasicHeaderApiTokenAuthenticator` will throw this up and prevent the API token from being used.

**Comment by jglick** [comment]
Created https://github.com/jenkinsci/ldap-plugin/pull/60 as a demonstration of what I mean, since it seems my earlier comments were not concrete enough.

Removing `ready-for-merge` since discussion is still active.

**Comment by jvz** [comment]
Ah, I see spring-security has a nifty new feature. That might work better. I'll look into that and see if I can close this PR.

**Comment by jglick** [comment]
It is not a new feature really (would have applied equally to Acegi Security as far as I know), just a different approach to the same problem.

**Comment by jvz** [comment]
Well, from what I recall, there were several caller sites that only catch `UsernameNotFoundException` and not the other four Acegi or Spring Security exception variants. Assuming an uncaught exception in those scenarios only leads to an ugly 500 page, that can be cleaned up as discovered.

**Comment by jvz** [comment]
Holding this as an alternative for now, though I think the alternate approach in https://github.com/jenkinsci/ldap-plugin/pull/61 should work well enough.

**Comment by jvz** [comment]
This PR is no longer needed due to changes in how we're handling this in the affected plugins directly.

**Pull request closed without merging**
